drivers/gpu/drm/xen/xen_drm_zcopy_drv.c

```diff
@@ -62,6 +99,112 @@ static inline struct xen_gem_object *to_xen_gem_obj(
 	return container_of(gem_obj, struct xen_gem_object, base);
 }
 
+static struct xen_wait_obj *xen_wait_obj_new(struct xen_drv_info *drv_info,
+	struct xen_gem_object *xen_obj)
+{
+	struct xen_wait_obj *wait_obj;
+
+	wait_obj = kzalloc(sizeof(*wait_obj), GFP_KERNEL);
+	if (!wait_obj)
+		return ERR_PTR(-ENOMEM);
+
+	init_completion(&wait_obj->completion);
+	wait_obj->xen_obj = xen_obj;
+	spin_lock(&drv_info->wait_list_lock);
+	list_add(&wait_obj->list, &drv_info->wait_obj_list);
+	spin_unlock(&drv_info->wait_list_lock);
+	return wait_obj;
+}
+
+static inline void xen_wait_obj_free(struct xen_drv_info *drv_info,
+	struct xen_wait_obj *wait_obj)
+{
+	struct xen_wait_obj *cur_wait_obj, *q;
+
+	spin_lock(&drv_info->wait_list_lock);
+	list_for_each_entry_safe(cur_wait_obj, q,
+			&drv_info->wait_obj_list, list) {
+		if (cur_wait_obj == wait_obj) {
+			list_del(&wait_obj->list);
+			kfree(wait_obj);
+			break;
+		}
+	}
+	spin_unlock(&drv_info->wait_list_lock);
+}
+
+static void xen_wait_obj_free_all(struct xen_drv_info *drv_info)
+{
+	struct xen_wait_obj *wait_obj, *q;
+
+	spin_lock(&drv_info->wait_list_lock);
+	list_for_each_entry_safe(wait_obj, q, &drv_info->wait_obj_list, list) {
+		complete_all(&wait_obj->completion);
+		list_del(&wait_obj->list);
```

Comment:
If you are calling `complete_all()` then you assume that client can be locked on a completion.
But then you are freeing `wait_obj`. ioctl code will do the same. This is not right.